src/transformers/parallelization_utils.py

```diff
@@ -0,0 +1,1098 @@
+
+            assert size_of_padding > 0, "pad size for vocab parallel embedding must be positive."
+        else:
+            size_of_padding = 0
+
+        for layer in self.policy.word_embedding(model, config):
+            if size_of_padding > 0:
+                padding = torch.zeros(
+                    size_of_padding,
+                    hidden_size,
+                    device=layer.weight.device,
+                    dtype=layer.weight.dtype,
+                )
+                layer.weight = torch.cat([layer.weight, padding], dim=0)
+
+            # embedding layer has (vocab_size, embedding_dims) tensor
+            # so dimension for chunking must be 0 for vocab parallel embedding
+            chunked_weight = torch.chunk(layer.weight, world_size, dim=0)
+            layer.weight.data = chunked_weight[gpu_index].to(self.device)
+            self._postprocess(layer, {"mpu": mpu})
+
+    def _process_repeated_layers(self, model, mpu, gpu_index, world_size, config):
+        for _, child in model.named_children():
+            if isinstance(child, self.policy.original_layer_class()):
+                self.policy.replace_arguments(
+                    layer=child,
+                    world_size=world_size,
+                    config=config,
+                )
+
+                parameters = [
+                    self._column_slice(
+                        self.policy.attn_qkv(child, config),
+                        world_size=world_size,
+                        gpu_index=gpu_index,
+                    ),
+                    self._row_slice(
+                        self.policy.attn_out(child, config),
+                        world_size=world_size,
+                        gpu_index=gpu_index,
+                    ),
+                    self._column_slice(
+                        self.policy.mlp_in(child, config),
+                        world_size=world_size,
+                        gpu_index=gpu_index,
+                    ),
+                    self._row_slice(
+                        self.policy.mlp_out(child, config),
+                        world_size=world_size,
+                        gpu_index=gpu_index,
+                    ),
+                ]
+
+                for layer in self.policy.layerwise_copy_to_all(child, config):
+                    if layer.weight is not None:
+                        layer.weight.data = layer.weight.to(self.device)
+                    if layer.bias is not None:
+                        layer.bias.data = layer.bias.to(self.device)
+                    parameters.append([layer])
+
+                for layers in parameters:
+                    for layer in layers:
+                        self._postprocess(
+                            layer,
+                            {
+                                "mpu": mpu,
+                                "skip_bias_add": self.skip_bias_add,
+                                "reversed": layer.reversed,
```

Comment:
It's for models like GPT2 and TransfoXL.
they used reversed weight (like Conv1D)

check this: https://github.com/huggingface/transformers/blob/1ed2ebf60d87ef12bd063c7c58e484e19189c754/src/transformers/modeling_utils.py#L1376